cudax/include/cuda/experimental/__device/attributes.cuh

```diff
@@ -0,0 +1,690 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of CUDA Experimental in CUDA C++ Core Libraries,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDAX__DEVICE_ATTRIBUTES_
+#define _CUDAX__DEVICE_ATTRIBUTES_
+
+#include <cuda_runtime_api.h>
+// cuda_runtime_api needs to come first
+
+#include <cuda/std/__cuda/api_wrapper.h>
+
+#include "cuda/experimental/__device/device.cuh"
+#include "cuda/std/__cccl/attributes.h"
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+namespace cuda::experimental
+{
+
+namespace detail
+{
+template <::cudaDeviceAttr Attr, typename Type>
+struct __attr_with_type
+{
+  using type = Type;
+  constexpr operator ::cudaDeviceAttr() const noexcept
+  {
+    return Attr;
+  }
+};
+} // namespace detail
+
+// TODO: give this a strong type for kilohertz
+template <>
+struct device::__attr<::cudaDevAttrClockRate> //
+    : detail::__attr_with_type<::cudaDevAttrClockRate, int>
+{};
+template <>
+struct device::__attr<::cudaDevAttrGpuOverlap> //
+    : detail::__attr_with_type<::cudaDevAttrGpuOverlap, bool>
+{};
+template <>
+struct device::__attr<::cudaDevAttrKernelExecTimeout> //
+    : detail::__attr_with_type<::cudaDevAttrKernelExecTimeout, bool>
+{};
+template <>
+struct device::__attr<::cudaDevAttrIntegrated> //
+    : detail::__attr_with_type<::cudaDevAttrIntegrated, bool>
+{};
+template <>
+struct device::__attr<::cudaDevAttrCanMapHostMemory> //
+    : detail::__attr_with_type<::cudaDevAttrCanMapHostMemory, bool>
+{};
+template <>
+struct device::__attr<::cudaDevAttrComputeMode> //
+    : detail::__attr_with_type<::cudaDevAttrComputeMode, ::cudaComputeMode>
+{
+  static constexpr type _default          = cudaComputeModeDefault;
```

Comment:
Nit/opinion: Would it make sense to redundantly call it `default_mode` instead of `_default`?

Reply:
i like that suggestion.